ctapipe/image/muon/processor.py

```diff
@@ -0,0 +1,211 @@
+            "(select) a given data value.  E.g. `[('mycut', 'x > 3'),]. "
+            "You may use `numpy` as `np` and `astropy.units` as `u`, but no other"
+            " modules."
+        ),
+    ).tag(config=True)
+
+
+class MuonProcessor(TelescopeComponent):
+
+    completeness_threshold = FloatTelescopeParameter(
+        default_value=30.0, help="Threshold for calculating the ``ring_completeness``"
+    ).tag(config=True)
+
+    ratio_width = FloatTelescopeParameter(
+        default_value=1.5,
+        help=(
+            "Ring width for intensity ratio"
+            " computation as multiple of pixel diameter"
+        ),
+    ).tag(config=True)
+
+    pedestal = FloatTelescopeParameter(
+        help="Pedestal noise rms", default_value=1.1
+    ).tag(config=True)
+
+    def __init__(self, subarray, **kwargs):
+        super().__init__(subarray, **kwargs)
+        self.dl1_query = DL1Query(parent=self)
+        self.ring_query = RingQuery(parent=self)
+        self.geometries = {
+            tel_id: tel.camera.geometry.transform_to(TelescopeFrame())
+            for tel_id, tel in self.subarray.tel.items()
+        }
+        self.fov_radius = {
+            tel_id: geometry.guess_radius()
+            for tel_id, geometry in self.geometries.items()
+        }
+
+        self.ring_fitter = MuonRingFitter(parent=self)
+
+        self.intensity_fitter = MuonIntensityFitter(subarray=subarray, parent=self)
+
+    def __call__(self, event: ArrayEventContainer):
+        for tel_id in event.dl1.tel:
+            self._process_telescope_event(event, tel_id)
+
+    def _process_telescope_event(self, event, tel_id):
+        event_index = event.index
+        event_id = event_index.event_id
+
+        if self.subarray.tel[tel_id].optics.n_mirrors != 1:
+            self.log.warn(
+                f"Skipping non-single mirror telescope {tel_id}"
+                " set --allowed_tels to get rid of this warning"
+            )
+            event.muon.tel[tel_id] = MuonCameraContainer()
+            return
+
+        self.log.debug(f"Processing event {event_id}, telescope {tel_id}")
+        dl1 = event.dl1.tel[tel_id]
+        image = dl1.image
+        mask = dl1.image_mask
+        if mask is None:
+            mask = image > 0
+
+        checks = self.dl1_query(p=dl1.parameters)
+
+        if not all(checks):
+            event.muon.tel[tel_id] = MuonCameraContainer()
+            return
+
+        geometry = self.geometries[tel_id]
+        x = geometry.pix_x
+        y = geometry.pix_y
+
+        # iterative ring fit.
+        # First use cleaning pixels, then only pixels close to the ring
+        # three iterations seems to be enough for most rings
+        for i in range(3):
+            ring = self.ring_fitter(x, y, image, mask)
+            dist = np.sqrt((x - ring.center_x) ** 2 + (y - ring.center_y) ** 2)
+            mask = np.abs(dist - ring.radius) / ring.radius < 0.4
```

Comment:
magic number? `0.4`

configurable?

Reply:
This line seems to be the cause of #740

Reply:
@moralejo

Reply:
Let's keep the behavior the same for now and just have the Tool refactoring changes. Another PR should be made for actual behavorial changes.